TANF Section 2 Validation

elipe17 opened the pull request:
## Summary of Changes
- Updated row schema's with cat2/3 validators
- Pulled in extra validators defined in 2369
Pull request closes #1109 

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```
- Log in as a user who can submit TANF data.
- Submit `small_tanf_section2.txt` and verify one T4 and one T5 record are created
- Submit `ADS.E2J.FTP2.TS06` and verify many T4 and T5 records are created. There will also be a lot of Field ParserErrors generated

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Data File parsing logic is in place such that data file submission can be completed by user and parsed by system for full submission once errors are resolved.
- [x] Data Errors tell user specifically where in the file they occur, and therefore what exactly needs to be fixed in order for the submission to be complete
- [x] TDP gives the user all data errors for a submission in a way they can easily save or export so they can take the issues back to their case management software and fix the issues then recreate an export for resubmission to TDP.
- [x]  `lfrohlich` and/or `adpennington` confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- ADding t2 schema

elipe17 committed:
- small fixes to naming
- added t2.py
- Updating var names to match field names

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- Added new doc for T2
- Updated model fo rT2
- Added active parse function

elipe17 committed:
- Added T3 schema defs

elipe17 committed:
- Fixing most lint errors

elipe17 committed:
- Updated T3 to multirow record

elipe17 committed:
- Added unit test and data files
- updated field names

elipe17 committed:
- Updating var names to conform to DB max var name length

elipe17 committed:
- Updating based on unit test failure

elipe17 committed:
- adding datafile

elipe17 committed:
- Adding unit tests for t2 and t3

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into feature/1113-1-ssp-section-1-models

elipe17 committed:
- Fixed field types
- Updated test file

elipe17 committed:
- Removing many migrations to consolodate into one migration

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
Revert "- Removing many migrations to consolodate into one migration"

This reverts commit 1455ae4d334f72e250405803d61d26c0a894e886.

elipe17 committed:
- Resolved test issues after merge

elipe17 committed:
- Resolve lint errors

elipe17 committed:
Merge branch 'feature/1113-1-ssp-section-1-models' of https://github.com/raft-tech/TANF-app into feature/1113-1-ssp-section-1-models

elipe17 committed:
Merge branch 'feature/1113-1-ssp-section-1-models' into 2424-ts1-parsing

elipe17 committed:
- Merged 1113.2

elipe17 committed:
Merge branch 'feature/1113-2-ssp-section-1-parsing' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- Removed unnecessary file

elipe17 committed:
- Rename model fields
- Combined migrations

elipe17 committed:
- fixed spelling in test

elipe17 committed:
- Fixed lint errors

jtimpe committed:
rm commented validators

elipe17 committed:
- Updated schema newlines to be consistent

ADPennington committed:
Merge branch 'develop' into feature/1113-2-ssp-section-1-parsing

elipe17 committed:
Merge branch 'feature/1113-2-ssp-section-1-parsing' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- Updated field names
- Updated tests
- Updated migration

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- consolodating migrations

elipe17 committed:
- Updated readme and backup script

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- Updated parse logic to batch create records

elipe17 committed:
- Fixed lint errors

elipe17 committed:
- Batching record serialization

elipe17 committed:
- Updated parse logic

elipe17 committed:
- Adding big s1 datafile

elipe17 committed:
- fixing lint errors

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance

elipe17 committed:
- Removing test to see if its causing weird failure

elipe17 committed:
- Updating based on comments
- Removing big file since circi cant handle it

elipe17 committed:
- fixing error in update method

elipe17 committed:
- fixing error in update

elipe17 committed:
- Removed extraneous seek

elipe17 committed:
- updating ignore to ignore patch files

elipe17 committed:
- Made evaluate_trailer pure/immutable

elipe17 committed:
Revert "- Removing test to see if its causing weird failure"

This reverts commit 64b78b737c97fb9bfa70217ff70ccffea8d85429.

elipe17 committed:
- Use custom transactions while parsing
- Handle transaction rollback on parse failure/error
- Update tests to use transaction logic
- Created new decorator to rollback database when assertion error is hit
- Making elastic search log level configurable
- Added test for inter parse rollback

elipe17 committed:
- updated store logic to one liner
- updated decorator to catch all exceptions
- updated tests

elipe17 committed:
- removed function
- renamed test

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance

elipe17 committed:
- refactored multirecord schema to be a schema manager instead
- updated parse functions to use the new layout
- updated bulk create to manage batch size

elipe17 committed:
- Name update for clarity

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance
- Fixed conflicts
- Updated tests to reflect new order of errors

elipe17 committed:
- Fix lint errors

elipe17 committed:
- Changing batch size to avoid crash in circi

elipe17 committed:
- Updated based on review
- Updated tests to reflect line number

elipe17 committed:
- Updated based on review/OH
- Added extra transaction safety

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance

elipe17 committed:
- Updated view to show latest datafiles
- Added admin filter to show newest or all datafile records
- Updated indices to allow easier elastic queries

elipe17 committed:
- Updated based on review  comments

elipe17 committed:
- Updated based on feedback for preparse error handling
- updated tests

elipe17 committed:
- Updated search indices to have parent FK

elipe17 committed:
- Fix lint errors

elipe17 committed:
- Updated submission tests
- Moved create_datafile to util

elipe17 committed:
- fix lint errors

elipe17 committed:
- removing frontend filtering

elipe17 committed:
- addding datafile to admin model

elipe17 committed:
Revert "- addding datafile to admin model"

This reverts commit 35a6f24c36c3a4c00ddcfc40f20833530b0199f4.

elipe17 committed:
- Fixed issue where datafile FK wasnt populating
- Regenerated migration

elipe17 committed:
- Readding datafile back to admin view now that the error is resolved

elipe17 committed:
- adding datafile back

elipe17 committed:
Revert "- Readding datafile back to admin view now that the error is resolved"

This reverts commit 2807425059fd1b5b355edfb16d30d170cf869d7b.

elipe17 committed:
- Removed unnecessary fields
- Updated dependencies
- Updated filter

elipe17 committed:
- Updated document to include required fields

elipe17 committed:
- Moved datafile reference to avoid confusion

ADPennington committed:
Merge branch 'develop' into 2486-parser-performance

elipe17 committed:
- Updating based on comments.
- Added base class to keep things a little more DRY

elipe17 committed:
Merge branch '2579-latest-data' into 2486-parser-performance

elipe17 committed:
- Refactored parsing to avoid custom transactions by leveraging the FK on the records. Rollbacks are a lot slower now, but hopefully that will happen much less versus successful parses.

elipe17 committed:
- small optimization for the rollback

elipe17 committed:
- Fix lint errors

elipe17 committed:
- Removing commented code

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2579-latest-data

elipe17 committed:
- Installing build essentials

elipe17 committed:
Revert "- Installing build essentials"

This reverts commit 782a5f7d687e60965a9100520f027b9678dbac35.

elipe17 committed:
- adding setup tools and wheel

elipe17 committed:
Revert "- adding setup tools and wheel"

This reverts commit f529728811fba242132b7c42f9e9e09d6037fa70.

elipe17 committed:
- Updating dependencies to get around pep issue

elipe17 committed:
- Pin factoryboy
- fix lint error

elipe17 committed:
Merge branch '2579-latest-data' into 2486-parser-performance

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance

elipe17 committed:
- Updating manifest

elipe17 committed:
- Added EncryptedField class
- Updated schema's
- Updated datafiles since all are encrypted

elipe17 committed:
- Fix lint errors

elipe17 committed:
- Added decryption for ssp

elipe17 committed:
- Making the encrypted check stronger

elipe17 committed:
- Added section two models, schemas, and an initial test

elipe17 committed:
- Allowing fields to be nullable

elipe17 committed:
- add missing field

elipe17 committed:
- Consolodating migrations

elipe17 committed:
- Consolodate migration
- Add tests

elipe17 committed:
Merge branch '332-ssn-encryption' into 1109-tanf-section2

elipe17 committed:
- Added encrypted field

elipe17 committed:
- Refactored fields and rowschema into seperate files
- Updated encrypted field to take decryption function

elipe17 committed:
- fix lint errors

elipe17 committed:
Merge branch '332-ssn-encryption' into 1109-tanf-section2

elipe17 committed:
- Fixed file spacing

elipe17 committed:
- fix imports
- fix lint errors

elipe17 committed:
- Fix import error

elipe17 committed:
- Fix failing test

elipe17 committed:
- Revert erroneous change

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1109-tanf-section2

elipe17 committed:
- Updating item numbers

elipe17 committed:
- Fixed item numbers

elipe17 committed:
- Updating to accomodate item number as string

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 332-ssn-encryption

elipe17 committed:
- Removed erroneous update that was causing error
- Fixed whitespace in datafiles

elipe17 committed:
- White space error

elipe17 committed:
- Resolved error

elipe17 committed:
Merge branch '332-ssn-encryption' into 1109-tanf-section2

elipe17 committed:
- Fixing test

elipe17 committed:
- fix lint errors

elipe17 committed:
- Added field validators for T4/T5

elipe17 committed:
- Added cat3 validators

elipe17 committed:
- Resolve lint and unit test errors

jtimpe commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t4.py:
```diff
@@ -20,27 +20,31 @@
               Field(item="6", name='CASE_NUMBER', type='string', startIndex=8, endIndex=19,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isAlphaNumeric(),
+                                               validators.notMatches('_'*11),
```
is the `_`-fill worth adding to `util.value_is_empty`? `value_is_empty` is used internally for `required` fields

elipe17 committed:
- small optimization to validator

jtimpe commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +15,128 @@
+          postparsing_validators=[
+              validators.if_then_validator(
+                  condition_field='FAMILY_AFFILIATION', condition_function=validators.notMatches(1),
+                  result_field='SSN', result_function=validators.isNumber()
```
the notes on this one leave me questioning the `result_function`

> if TANF recipient, then cannot have all 0s,1s,2s,3s,4s,5s,6s,7s,8s,or 9s for SSN.

#2518 handles it more like i'd expect
https://github.com/raft-tech/TANF-app/blob/4b70e6aa36b0cc605579d38e6eabf679fe3811b7/tdrs-backend/tdpservice/parsers/schema_defs/tanf/t2.py#L18

jtimpe commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +15,128 @@
+                  ),
+              validators.validate__FAM_AFF__HOH__FEDTIME(),
+              validators.if_then_validator(
+                  condition_field='DATE_OF_BIRTH', condition_function=validators.month_year_yearIsLargerThan(19),
```
i think `month_year_yearIsLargerThan` checks the first four characters of the string (the year portion), whereas the notes seem to indicate that we should calculate the age of the recipient 

> adults must have a valid value for 19A; requires calculating AGE: DOB (item #15) must be valid; convert RPT_MON_YEAR (item #4) to YYYYMMDD and set DD to 01; AGE formula: (rpt_month_year - DOB)/365.25

andrew-jameson replied:
Can't datetime simply ingest such a format with strftime or a function like that? Should make this a couple lines.

Not a fan of the function name mixing camelCase and underscores. Can we simplify it down to `yearIsLargerThan()`?

elipe17 committed:
- Added tests for cat3 validators for T5 record

elipe17 committed:
- Fix lint errors

jtimpe commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +15,128 @@
+                  condition_field='FAMILY_AFFILIATION', condition_function=validators.notMatches(1),
+                  result_field='SSN', result_function=validators.isNumber()
+                  ),
+              validators.validate__FAM_AFF__SSN(),
```
curious if you played with the `and_validator` introduced in #2518 
https://github.com/raft-tech/TANF-app/blob/4b70e6aa36b0cc605579d38e6eabf679fe3811b7/tdrs-backend/tdpservice/parsers/validators.py#L22
i think that may give us enough flexibility to move away from custom validator functions in most cases. open to opinions though as i'm sure the expressions could get gnarly. i'd be a fan of at least utilizing the `make_validator` from the custom function so that the functions are consistent.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +15,128 @@
                     required=True, validators=[]),
               Field(item="4", name='RPT_MONTH_YEAR', type='number', startIndex=2, endIndex=8,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.month_year_yearIsLargerThan(1998),
```
Why is 1998 hardcoded here?

elipe17 committed:
- updated fields to correct types

elipe17 committed:
- update race/ethnicity

elipe17 committed:
Merge branch '1109-tanf-section2' into 1109-tanf-section2-validation

elipe17 committed:
- updated tests post merge
- added check in make_validator to just return false if the value is None

elipe17 committed:
- Updated based on review feedback

elipe17 committed:
- Fix lint error

elipe17 committed:
- Updating to the correct types

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1109-tanf-section2

elipe17 committed:
- Resolve conflicts

elipe17 committed:
- Moving field updates to seperate migration to see if they take place

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1109-tanf-section2

elipe17 committed:
Revert "- Moving field updates to seperate migration to see if they take place"

This reverts commit a3214f23e0bd9cc4cc37ad40e63ae0c98b4ea64b.

elipe17 committed:
- Revert migration 15 to original state
- Generate new migration for field alterations

elipe17 committed:
Merge branch '1109-tanf-section2' into 1109-tanf-section2-validation

elipe17 committed:
- Merged base branch and updated tests/factories

elipe17 committed:
- Fixed test

elipe17 committed:
- Fixed field names to correspond to model

elipe17 committed:
Merge branch '1109-tanf-section2' into 1109-tanf-section2-validation

andrew-jameson commented on tdrs-backend/tdpservice/parsers/validators.py:
```diff
@@ -61,16 +129,153 @@ def startsWith(substring):
 
 # custom validators
 
+def month_year_monthIsValid():
+    """Validate that in a monthyear combination, the month is a valid month."""
+    return make_validator(
+        lambda value: int(str(value)[4:6]) in range(1, 13),
+        lambda value: f'{str(value)[4:6]} is not a valid month.'
+    )
+
+def olderThan(min_age):
+    """Validate that value is larger than min_age."""
+    return make_validator(
+        lambda value: date.today().year - int(str(value)[:4]) > min_age,
+        lambda value: f'{date.today().year - int(str(value)[:4])} is not larger than {min_age}.'
+    )
+
+def month_year_yearIsLargerThan(year):
```
Can we improve these function names with the following forms?
`dateMonthIsValid()`, `dateOlderThan()`, and `dateYearIsLargerThan()`?

elipe17 replied:
Do you want that change made here or in the 2369 PR? This function is from 2369.

andrew-jameson replied:
My apologies. Please do make the change here as 2369 is deep into qasp and alex did not want any new commits. I'm just not a fan of combining underscores and camelCase.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -22,6 +22,17 @@ def create_test_datafile(filename, stt_user, stt, section='Active Case Data'):
     return datafile
 
 
+def value_is_empty(value, length):
+    """Handle 'empty' values as field inputs."""
+    empty_values = [
+        ' '*length,  # '     '
+        '#'*length,  # '#####'
+        '_'*length,  # '_____'
+    ]
+
+    return value is None or value in empty_values
+
```
I just deleted this snippet in 1613 merge. Is this actually needed?

andrew-jameson commented on tdrs-backend/tdpservice/parsers/test/test_parse.py:
```diff
@@ -631,7 +631,8 @@ def test_parse_tanf_section2_file(tanf_section2_file):
 
     parser_errors = ParserError.objects.filter(file=tanf_section2_file)
-    assert parser_errors.count() == 2681
+    assert parser_errors.count() == 3067
```
Per @raftmsohani , I think we should remove these asserts with some large error count.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +15,131 @@
+          postparsing_validators=[
+              validators.if_then_validator(
+                  condition_field='FAMILY_AFFILIATION', condition_function=validators.notMatches(1),
+                  result_field='SSN', result_function=validators.notOneOf(("000000000", "111111111", "222222222",
```
I may have said this earlier, can we just make a reusable SSN validator to prevent this repetition?

elipe17 replied:
If we use the DRY idiom wrt the validators then we should be making specific validators for race/ethnicity, earned income fields, and basically every field in the T6 record to avoid duplicating the same validator. Im good with doing that, I just want to keep it consistent. Let me know how you'd like me to proceed.

andrew-jameson replied:
The repetition is moreso the silly-looking "00000" stuff which takes up a few lines but is repeated frequently across the RowSchemas. I'd like the SSN to have a standalone and cleaner-looking validator here. 

If we feel that there are other areas to optimize which improve readability and code smell, I'm not opposed to updating those if it's not a large lift. So, soft yes if you feel we are repeating the race/ethnicity validation.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/test/test_validators.py:
```diff
@@ -1,6 +1,8 @@
 
+import pytest
 from .. import validators
+from tdpservice.parsers.test.factories import TanfT5Factory
```
We make a T4 factory but don't need to use it? Maybe I missed it in test_parse?

elipe17 replied:
Dont use it currently. I can remove if you want. I kept it for continuity and if there happen to be cat3 validators added for t4.

andrew-jameson replied:
Hmm, I hesitate to include unused/untested code in a PR. Let's remove it for now but ensure to not delete this branch when merging in case we want to reference it later.

andrew-jameson commented:
Submitted a different file up to tanf section 2, no records seem to have been created?
```
   2023-08-08T11:23:01.70-0400 [APP/PROC/WEB/0] ERR [2023-08-08 15:23:01,707 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 0 errors: {}
   2023-08-08T11:23:01.70-0400 [APP/PROC/WEB/0] ERR [2023-08-08 15:23:01,707: INFO/ForkPoolWorker-1] DataFile parsing finished with 0 errors: {}
   ```
   
   
![Screenshot 2023-08-08 at 11 29 53 AM](https://github.com/raft-tech/TANF-app/assets/84722778/124eca35-0aa9-4998-8d0a-d79059f51190)
![Screenshot 2023-08-08 at 11 29 02 AM](https://github.com/raft-tech/TANF-app/assets/84722778/8a9d6964-915b-489d-bf4c-36f04c4f1531)

But all t4/t5 objects are only associated to older datafiles -- nothing appears to have been generated for datafile 17.

elipe17 commented:
> Submitted a different file up to tanf section 2, no records seem to have been created?
> 
> ```
>    2023-08-08T11:23:01.70-0400 [APP/PROC/WEB/0] ERR [2023-08-08 15:23:01,707 INFO parser_task.py::parse:L21 :  DataFile parsing finished with 0 errors: {}
>    2023-08-08T11:23:01.70-0400 [APP/PROC/WEB/0] ERR [2023-08-08 15:23:01,707: INFO/ForkPoolWorker-1] DataFile parsing finished with 0 errors: {}
> ```
> 
> ![Screenshot 2023-08-08 at 11 29 53 AM](https://user-images.githubusercontent.com/84722778/259150470-124eca35-0aa9-4998-8d0a-d79059f51190.png) ![Screenshot 2023-08-08 at 11 29 02 AM](https://user-images.githubusercontent.com/84722778/259150517-8a9d6964-915b-489d-bf4c-36f04c4f1531.png)
> 
> But all t4/t5 objects are only associated to older datafiles -- nothing appears to have been generated for datafile 17.

Did you submit the two datafiles shown at the same time?

elipe17 committed:
- Remove duplicate function
- Update ssn validator
- Removed asserts with large number of parse errors

raftmsohani reviewed (approved):
LGTM!

elipe17 committed:
- Remove print statemtnt

elipe17 committed:
- Resolved filter issue that made it seem like records werent being created
- Added filter to remaining records

elipe17 committed:
- fix lint errors

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1109-tanf-section2-validation

elipe17 committed:
- correct form merge conflict

elipe17 commented:
@andrew-jameson  When you have some time could you check my responses to your comments and let me know how you would like me to proceed?

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1109-tanf-section2-validation

andrew-jameson commented:
> > Submitted a different file up to tanf section 2, no records seem to have been created?
> > But all t4/t5 objects are only associated to older datafiles -- nothing appears to have been generated for datafile 17.
> 
> Did you submit the two datafiles shown at the same time?

I think I may have, yes. Why should that matter?

elipe17 committed:
- Added specific validators to avoid duplication
- Updated naming of validators

andrew-jameson reviewed (approved):
These little tweaks I feel made a huge difference! Looks great, thank you for that.

andrew-jameson commented:
> These little tweaks I feel made a huge difference! Looks great, thank you for that.

For posteriority -- there was a bug when submitting multiple datafiles at once that was resolved elsewhere and merged in during review of this PR. Resolved.

elipe17 commented:
> > These little tweaks I feel made a huge difference! Looks great, thank you for that.
> 
> For posteriority -- there was a bug when submitting multiple datafiles at once that was resolved elsewhere and merged in during review of this PR. Resolved.

Just clarifying that the bug was resolved in this PR in `filter.py`.

elipe17 committed:
TANF Section 3 Parsing/Validation (#2649)

* - Updated to support parsing section 3 data

* - Added validators and tests for T6 record

* - resolve lint errors

* - Quick rename

* - Update fields in test

* - Fixed conflict

* - Fix error from merge conflict

* - Updated to create RPT_MONTH_YEAR from CALENDAR_QUARTER
- Updated tests

* - Added new validators for field change

* - Remove debug code

* - Genericize TransformField
- Update all schemas to use TransformField
- Move transforms to seperate file
- Fix lint errors

* - put kwargs in correct spot

* - Very minor change to avoid setting field unnecessarily on all TransformFields

* - Updated validator to call out name of field versus the value
- Updated tests

* - Fix lint

elipe17 committed:
- Calling super to avoid duplicate code

elipe17 committed:
- Added validators for transformed field

ADPennington commented:
@elipe17 can you resolve conflicts and let me know when this is ready?

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1109-tanf-section2-validation

elipe17 committed:
- Updating based on merge

elipe17 committed:
- Fixed remaining merge conflicts

elipe17 committed:
- Fix lint errors

elipe17 committed:
- update error messages based on validator updates

ADPennington committed:
Merge branch 'develop' into 1109-tanf-section2-validation

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,129 @@
-          postparsing_validators=[],
+          postparsing_validators=[
+              validators.if_then_validator(
+                  condition_field='FAMILY_AFFILIATION', condition_function=validators.notMatches(1),
```
@elipe17 where is the following validator?

`If item 14 ==1 then item 16 != 000000000 -- 999999999`

elipe17 replied:
Should have been `matches` instead of `notMatches`

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,129 @@
-              EncryptedField(decryption_func=tanf_ssn_decryption_func, item="16", name='SSN', type='string',
-                             startIndex=28, endIndex=37, required=True, validators=[]),
+                    required=True, validators=[validators.isLargerThan(0)]),
+              TransformField(transform_func=tanf_ssn_decryption_func, item="16", name='SSN', type='string',
```
@elipe17 we are seeing evidence of SSNs being decrypted when encryption indicator is not `E`. see below:

![2633p1](https://github.com/raft-tech/TANF-app/assets/63075587/d5541513-0bfe-4227-b579-8b0fcd635468)

elipe17 replied:
@ADPennington I am going to try redeploying. I am not able to reproduce this locally. Screenshots below of what I used to test. Could you send me your test file if the issue persists to see if I can reproduce the issue?
![Screenshot 2023-08-28 at 3 34 24 PM](https://github.com/raft-tech/TANF-app/assets/125676261/42b709f3-cacf-4e8f-b64e-7d5256942320)

elipe17 replied:
@ADPennington update. I redeployed to qasp environment and performed the same test and saw the same result. Could you try again and see if that was the issue for you too?
![Screenshot 2023-08-28 at 4 21 52 PM](https://github.com/raft-tech/TANF-app/assets/125676261/89e79e55-13fb-465b-b8d4-6118baeb3110)

ADPennington replied:
@elipe17 still seeing it
![2633p3](https://github.com/raft-tech/TANF-app/assets/63075587/b2ca5c46-af29-4050-aa0f-6c0d299c7d85)

elipe17 replied:
@ADPennington could you share your datafile with me so that I can debug with it?

elipe17 committed:
- Corrected validator

elipe17 committed:
Merge branch '1109-tanf-section2-validation' of https://github.com/raft-tech/TANF-app into 1109-tanf-section2-validation

elipe17 committed:
- Resolved bug causing file to think it was encrytped after multiple submissions and changing the encryption header in between submissions

ADPennington committed:
Merge branch 'develop' into 1109-tanf-section2-validation

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t4.py:
```diff
@@ -20,27 +20,30 @@
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isAlphaNumeric(),
+                                               validators.notEmpty()]),
               Field(item="2", name='COUNTY_FIPS_CODE', type='string', startIndex=19, endIndex=22,
```
@elipe17 im curious why we use a different validator here than we use in section 1 (see below):

```
 Field(item="2", name='COUNTY_FIPS_CODE', type='string', startIndex=19, endIndex=22,
                        required=True, validators=[
                              validators.isNumber(),
                        ]),
```

elipe17 replied:
I probably copied it from section 1 prior to it being changed in section 1. Will update to be consistent.

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t4.py:
```diff
@@ -20,27 +20,30 @@
+                    required=True, validators=[validators.isInStringRange(1, 999)]),
               Field(item="5", name='STRATUM', type='string', startIndex=22, endIndex=24,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInStringRange(1, 99)]),
```
```suggestion
                    required=True, validators=[validators.isInStringRange(0, 99)]),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t4.py:
```diff
@@ -20,27 +20,30 @@
+                                               validators.dateMonthIsValid()]),
               Field(item="6", name='CASE_NUMBER', type='string', startIndex=8, endIndex=19,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isAlphaNumeric(),
```
@elipe17 here too, i think we'd like it to be consistent with the validators in section 1. whichever is most complete. 

```
Field(item="6", name='CASE_NUMBER', type='string', startIndex=8, endIndex=19,
                        required=True, validators=[
                              validators.isAlphaNumeric(),
                              validators.notMatches('_'*11),
                              validators.notEmpty(),
                        ]),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,129 @@
+              validators.validate__FAM_AFF__SSN(),
+              validators.if_then_validator(
+                  condition_field='FAMILY_AFFILIATION', condition_function=validators.isInLimits(1, 3),
+                  result_field='RACE_HISPANIC', result_function=validators.validateRace()
```
@elipe17 i think these race-related checks are slightly incorrect. looks like the validator allows for value == 0 which is a valid possible response, but not in the scenario where fam affil == 1, 2, or 3. from section 1:

```
     validators.if_then_validator(
                    condition_field='FAMILY_AFFILIATION', condition_function=validators.isInLimits(1, 3),
                    result_field='RACE_HISPANIC', result_function=validators.oneOf((1, 2)),
                ),
```

ADPennington replied:
@elipe17 i think this may still need review. 0 is not a valid response for this condition. 

```
def validateRace():
    """Validate race."""
    return make_validator(
        lambda value: value >= 0 and value <= 2,
        lambda value: f'{value} is not greater than or equal to 0 or smaller than or equal to 1.'
```

elipe17 committed:
- Made case number validators consistent

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                  ),
+              validators.if_then_validator(
+                  condition_field='FAMILY_AFFILIATION', condition_function=validators.isInLimits(1, 3),
+                  result_field='MARITAL_STATUS', result_function=validators.isInLimits(0, 5)
```
```suggestion
                  result_field='MARITAL_STATUS', result_function=validators.isInLimits(1, 5)
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                  ),
+              validators.if_then_validator(
+                  condition_field='FAMILY_AFFILIATION', condition_function=validators.isInLimits(1, 3),
+                  result_field='EDUCATION_LEVEL', result_function=validators.or_validators(
```
@elipe17 from section 1:

```
validators.if_then_validator(
                        condition_field='FAMILY_AFFILIATION', condition_function=validators.isInLimits(1, 3),
                        result_field='EDUCATION_LEVEL', result_function=validators.or_validators(
                            validators.isInStringRange(0, 16),
                            validators.isInStringRange(98, 99)
                        ),
                    ),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/validators.py:
```diff
@@ -225,24 +239,52 @@ def isInLimits(LowerBound, UpperBound):
         lambda value: f'{str(value)[4:6]} is not a valid month.'
     )
 
+def olderThan(min_age):
```
@elipe17 age is calculated by taking the difference between rpt month  year and DOB:
```
requires calculating AGE: DOB (item #15) must be valid; convert RPT_MON_YEAR (item #4) to YYYYMMDD and set DD to 01; AGE formula: (rpt_month_year - DOB)/365.25  
```
i  believe @andrew-jameson mentioned that this would be a different ticket?

elipe17 replied:
Ill take a look at how big the lift would be to do this. If it is pretty low Ill just do it on this ticket.

elipe17 replied:
Ok, technically I could implement this with reasonable ease... But It will break some meaningful paradigms that we dont want to break. With that said, this would be easily implemented with the ticket (may or may not exist) that will refactor our schema creation so that we don't break these paradigms. I can explain in more detail if you would like, just let me know.

ADPennington replied:
added to #2688

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                  ),
+              validators.validate__FAM_AFF__HOH__Count_Fed_Time(),
+              validators.if_then_validator(
+                  condition_field='DATE_OF_BIRTH', condition_function=validators.olderThan(18),
```
this validator `olderThan` will need to be updated @elipe17

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                                               validators.dateMonthIsValid()]),
               Field(item="6", name='CASE_NUMBER', type='string', startIndex=8, endIndex=19,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isAlphaNumeric()]),
```
same comment as above re: `CASE_NUMBER`

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                             is_encrypted=False),
               Field(item="17A", name='RACE_HISPANIC', type='number', startIndex=37, endIndex=38,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInLimits(1, 3)]),
```
```suggestion
                    required=True, validators=[validators.isInLimits(0, 3)]),
```

all of the race/ethnicity fields can have a value of `0`

elipe17 committed:
- Converging validators between sections

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                    required=True, validators=[validators.isInLimits(0, 9)]),
               Field(item="19A", name='REC_OASDI_INSURANCE', type='number', startIndex=44, endIndex=45,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInLimits(1, 3)]),
```
```suggestion
                    required=True, validators=[validators.isInLimits(0, 2)]),
```

ADPennington replied:
this is almost ready to merge when this gets resolved @elipe17

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                    required=True, validators=[validators.isInLimits(1, 3)]),
               Field(item="19D", name='REC_AID_AGED_BLIND', type='number', startIndex=47, endIndex=48,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInLimits(1, 3)]),
```
```suggestion
                    required=True, validators=[validators.isInLimits(0, 2)]),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                    required=True, validators=[validators.isInLimits(1, 2)]),
               Field(item="19C", name='REC_AID_TOTALLY_DISABLED', type='number', startIndex=46, endIndex=47,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInLimits(1, 3)]),
```
```suggestion
                    required=True, validators=[validators.isInLimits(0, 2)]),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                    required=True, validators=[validators.isInLimits(1, 3)]),
               Field(item="19E", name='REC_SSI', type='number', startIndex=48, endIndex=49,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInLimits(1, 3)]),
```
```suggestion
                    required=True, validators=[validators.isInLimits(1, 2)]),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                    required=True, validators=[validators.isInStringRange(1, 10)]),
               Field(item="22", name='PARENT_MINOR_CHILD', type='number', startIndex=52, endIndex=53,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInLimits(1, 3)]),
```
```suggestion
                    required=True, validators=[validators.isInLimits(0, 2)]),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                    required=True, validators=[validators.isInLimits(1, 3)]),
               Field(item="23", name='NEEDS_OF_PREGNANT_WOMAN', type='number', startIndex=53, endIndex=54,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInLimits(1, 9)]),
```
```suggestion
                    required=True, validators=[validators.isInLimits(0, 9)]),
```

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                    required=True, validators=[validators.isInLimits(1, 9)]),
               Field(item="24", name='EDUCATION_LEVEL', type='string', startIndex=54, endIndex=56,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.or_validators(validators.isInStringRange(0, 16, 2),
```
what's the 2 mean here?

elipe17 replied:
Was for zfill, but that is no longer required in the latest updates.

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                                                                        validators.matches(9))]),
               Field(item="26", name='COUNTABLE_MONTH_FED_TIME', type='string', startIndex=57, endIndex=60,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInStringRange(0, 999, 3)]),
```
does 3 mean 3 digits?

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                    required=True, validators=[validators.isInStringRange(0, 999, 3)]),
               Field(item="27", name='COUNTABLE_MONTHS_STATE_TRIBE', type='string', startIndex=60, endIndex=62,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInStringRange(0, 99, 2)]),
```
same as above re: 2. 2 digits? is this a new parameter we're using?

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t5.py:
```diff
@@ -15,66 +16,128 @@
+                    required=True, validators=[validators.isInStringRange(0, 9999, 4)]),
               Field(item="30", name='AMOUNT_UNEARNED_INCOME', type='string', startIndex=67, endIndex=71,
-                    required=True, validators=[]),
+                    required=True, validators=[validators.isInStringRange(0, 9999, 4)]),
```
ditto above re: 4 digits?

ADPennington commented on tdrs-backend/tdpservice/parsers/test/factories.py:
```diff
@@ -193,3 +193,88 @@ class Meta:
+    STRATUM = 1
+    ZIP_CODE = "11111"
+    DISPOSITION = 1
+    CLOSURE_REASON = 1
```
can be 2 digits with leading zero

ADPennington reviewed (changes requested):
@elipe17 a few minor updates requested.

elipe17 committed:
- Updating based on review comments

elipe17 committed:
- Fix paren error

elipe17 committed:
- Removed extra dot

elipe17 committed:
- Updating validator

elipe17 committed:
- updating race conditional validators

elipe17 commented:
@ADPennington I updated the race conditional validators. Sorry about that. I also responded to the remaining un-resolved comment you had. Let me know what you think.

elipe17 committed:
- corrected validator

ADPennington reviewed (approved):
thanks @elipe17 🚀 


as noted below, created #2688 to track the follow-on work to wrap up tanf section 2 validation cc: @Smithh-Co
